tests/pass/libc.rs

```diff
@@ -5,13 +5,94 @@
         .map(std::path::PathBuf::from)
         .unwrap_or_else(|_| std::env::temp_dir())
 }
 
+/// Test allocating variant of `realpath`.
+fn test_posix_realpath_alloc() {
+    use std::ffi::OsString;
+    use std::ffi::{CStr, CString};
+    use std::fs::{remove_file, File};
+    use std::os::unix::ffi::OsStrExt;
+    use std::os::unix::ffi::OsStringExt;
+    use std::path::PathBuf;
+
+    let buf;
+    let path = tmp().join("miri_test_libc_posix_realpath_alloc");
+    let c_path = CString::new(path.as_os_str().as_bytes()).expect("CString::new failed");
+
+    // Cleanup before test.
+    remove_file(&path).ok();
+    // Create file.
+    drop(File::create(&path).unwrap());
+    unsafe {
+        let r = libc::realpath(c_path.as_ptr(), std::ptr::null_mut());
+        assert!(!r.is_null());
+        buf = CStr::from_ptr(r).to_bytes().to_vec();
+        libc::free(r as *mut _);
+    }
+    let canonical = PathBuf::from(OsString::from_vec(buf));
+    // Cleanup after test.
+    remove_file(&path).unwrap();
+
+    assert_eq!(path.file_name(), canonical.file_name());
+}
+
+/// Test non-allocating variant of `realpath`.
+fn test_posix_realpath_noalloc() {
+    use std::ffi::{CStr, CString};
+    use std::fs::{remove_file, File};
+    use std::os::unix::ffi::OsStrExt;
+    use std::path::PathBuf;
+
+    let path = tmp().join("miri_test_libc_posix_realpath_noalloc");
+    let c_path = CString::new(path.as_os_str().as_bytes()).expect("CString::new failed");
+
+    let mut v = vec![0; (libc::PATH_MAX + 1) as usize];
+
+    // Cleanup before test.
+    remove_file(&path).ok();
+    // Create file.
+    drop(File::create(&path).unwrap());
+    unsafe {
+        let r = libc::realpath(c_path.as_ptr(), v.as_mut_ptr());
+        assert!(!r.is_null());
+    }
+    let c = unsafe { CStr::from_ptr(v.as_ptr()) };
+    let canonical = PathBuf::from(c.to_str().expect("CStr to str"));
+    // Cleanup after test.
+    remove_file(&path).unwrap();
+
+    assert_eq!(path.file_name(), canonical.file_name());
+}
+
+/// Test failure cases for `realpath`.
+fn test_posix_realpath_errors() {
+    use std::convert::TryInto;
+    use std::ffi::CString;
+    use std::io::ErrorKind;
+
+    // Test non-existent path returns an error.
+    let c_path = CString::new("./nothing_to_see_here").expect("CString::new failed");
+    let r = unsafe { libc::realpath(c_path.as_ptr(), std::ptr::null_mut()) };
+    assert!(r.is_null());
+    let e = std::io::Error::last_os_error();
+    assert_eq!(e.raw_os_error(), Some(libc::ENOENT));
+    assert_eq!(e.kind(), ErrorKind::NotFound);
+
+    // Test long path returns an error.
+    let too_long = "x/".repeat(libc::PATH_MAX.try_into().unwrap());
+    let c_path = CString::new(too_long.as_bytes()).expect("CString::new failed");
+    let r = unsafe { libc::realpath(c_path.as_ptr(), std::ptr::null_mut()) };
+    assert!(r.is_null());
+    let e = std::io::Error::last_os_error();
+
+    assert!(e.raw_os_error().is_some());
```

Comment:
I meant to call this out as I am a bit turned around. I can't check because it is experimental: https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidFilename

Miri has it available because the feature is enabled for the compiler (I guess?). So calling `canonicalize` from Miri returns it as an option. I tried to assert here with `cfg` but I couldn't get the right incarnation. As-is I think we are exposing `InvalidFilename` to stable code that can't check it, which is probably wrong. Would love advice here.

Reply:
Just add a `#[feature]` flag to the top of the test.

Reply:
Done.